### ISSUE TITLE
Variable assignment highlighting

### ISSUE THREAD

**Issue by dgutov** [pedz/ruby-ts-mode#3]
`font-lock-doc-face` doesn't seem ideal -- e.g. it looks like a dull grey in my current theme.

The built-in face `font-lock-variable-name-face` seems more pertinent, and it's already used this way in `js-ts-mode` and `c-ts-mode`.

Example from there:

```el
   :language 'javascript
   :override t
   :feature 'expression
   `((assignment_expression
      left: [(identifier) @font-lock-function-name-face
             (member_expression property: (property_identifier)
                                @font-lock-function-name-face)]
      right: [(function) (arrow_function)])

     (call_expression
      function: [(identifier) @font-lock-function-name-face
                 (member_expression
                  property:
                  (property_identifier) @font-lock-function-name-face)])

     (assignment_expression
      left: [(identifier) @font-lock-variable-name-face
             (member_expression
              property: (property_identifier) @font-lock-variable-name-face)]))
```

Additionally, the instance and global variables (`$...` and `@...`) don't seem to be highlighted here at all.

![Screenshot from 2022-12-27 01-08-10](https://user-images.githubusercontent.com/271877/209587735-96faef00-93cc-476a-9bee-27cb6e190696.png)

**Comment by dgutov** [comment]
I dropped assignment highlighting when merging because it seemed less useful than simply highlighting the instance and global variables. And because we don't have separate faces for such expressions (the core recommends reusing `font-lock-variable-name-face`).

Hopefully we'll get the "locals highlighting" soon enough, and we can spend the "angry fruit salad" budget on that.

Please speak up if you disagree, of course.

**Comment by pedz** [comment]
> Hopefully we'll get the "locals highlighting" soon enough, and we can spend the "angry fruit salad" budget on that.

Sorry, I don't understand either of the quotes.

I'm leaning towards forking a branch.  I understand why it is important for the FSF code to be consistent but one of my main reasons for diving into this is because I've never agreed with much of their styles, indentation, etc.  I tend to agree only about 50% with Rubocop.  I don't know if I have ever used the standard ruby-mode.  Nor do I really use c-mode's indentation.  I've been using Emacs since it was first released in 1985 but there are many things that I don't like and either don't use of create my own versions.

I don't see myself as a follower.  I've been consulting for IBM and other companies my entire career and I assume I've encountered over 1000 programmers but I can count on my hands the number that use Emacs.  I would say that less than 1% of programmers use Emacs and this is within Unix type programmers.  Likewise for Ruby.  I know zero people that use Ruby -- which stuns me whenever I think about it.  The net is I have no interest to indent or font lock like others because I have no others.

I really do understand why it is important for the two modes to be consistent if / when it is part of the official distribution.  But it isn't the direction I want to end up in.

Right now, I'm spending time looking for hotels in Socorro, NM.  I'm in central TX.  Socorro has two big wildlife refuges and I've promised myself to go back there around Jan. 15th of this year.  [It will have been 20 years since my last trip there](https://www.behance.net/gallery/22063357/Bosque-del-Apache-2001).  I do amateur photography and my aim in retirement is to just drift around piddling, taking pictures, etc.  I'm sure I will start back piddling with Emacs later today or tomorrow -- although I've left another project I call "Hatred" unattended for some time and I might piddle with it.  It is a replacement for Quicken / Quick Books.

I'm toying with implementing a def-feature which would make it simpler to create a font lock feature, add it to the `treesit-font-lock-feature-list` list, define a face, etc.  Rather than list them in levels, just have a user option with them all listed and a checkbox beside each.  When customized completes, a call to `treesit-font-lock-recompute-features` is done.

For me, highlighting the l-values is much more important than differentiating between local, global, and instance variables.  In debugging, it is the places that muck with the values that you need to find -- not the references.  There are also concepts which appear to have been pushed off to lsp-mode / elgot which seem misplaced.

Your suggestion of making the indent rules look at the appropriate option at indent time was a huge game changer for me.    Rather than a "style" like c-mode has a bevy of individual options seem more enticing.  I wish the font lock features were as dynamic.  I can imagine adding commands that would turn particular highlighting on or off or even make some particular font lock faces blink briefly.  But the way the lists needs to be recomputed seems fragile to me.  It also seems like faces should be buffer local which I guess people have managed to do but that isn't the default out of the box case.

Anyhow... that's my thoughts right now.

**Comment by dgutov** [comment]
> Sorry, I don't understand either of the quotes.

The "locals" is the feature described in this section here: https://tree-sitter.github.io/tree-sitter/syntax-highlighting#local-variables

See the code box with the example. But this will require additional infrastructure for scanning and storing the scopes while going through the buffer.

And "angry fruit salad" is something people call code buffers/files/views with excessive syntax highlighting. When, subjectively, it crosses the border from being useful into just too distracting.

> I really do understand why it is important for the two modes to be consistent if / when it is part of the official distribution. But it isn't the direction I want to end up in.

FWIW, nobody officially asked me for this, it was my idea. But it does seem to make sense, to make things easier for the users, and to simplify the test harness as well. I do know people who write Ruby, though, both online and offline.

Is there something in particular that gives you a pause? Do you strongly disagree with the defaults?

I can understand if you don't want the overhead of coordinating with others (myself, or the treesit.el maintainer), or having to follow the arcane contributing rules.

But if it's not that, or about the defaults, we can always add support for additional indentation styles, or font-lock features. E.g. we could add the highlighting for assignments, but put them on the feature level 4.

> I wish the font lock features were as dynamic. I can imagine adding commands that would turn particular highlighting on or off or even make some particular font lock faces blink briefly. But the way the lists needs to be recomputed seems fragile to me.

I agree that there doesn't seem to be a hard technical reason why the feature list must be recomputed instead of following the level value at runtime. Perhaps this improves the performance a little, but I kind of doubt that the difference can be significant.

> It also seems like faces should be buffer local which I guess people have managed to do but that isn't the default out of the box case.

I'm not sure what you mean by buffer-local faces.

The Bosque del Apache shots from 2001 look great. I hope to have a similar kind of retirement one day.

**Comment by pedz** [comment]
> But this will require additional infrastructure for scanning and storing the scopes while going through the buffer.

Can you elaborate on this?  Is this extra C code in someplace like treesit.c?  Or extra lisp code in treesit.el?  Or extra code in the modes like ruby-ts-mode.el?

On a side topic, I like the use of underline in the example you pointed to.

> FWIW, nobody officially asked me for this, it was my idea. But it does seem to make sense, to make things easier for the users, and to simplify the test harness as well. I do know people who write Ruby, though, both online and offline.
> < snip >
> But if it's not that, or about the defaults, we can always add support for additional indentation styles, or font-lock features. E.g. we could add the highlighting for assignments, but put them on the feature level 4.

Mostly it is the indent rules that give me pause but I also get the feeling of being tied to a paradigm that is now out dated and also extremely Lisp focused.   Also, I assume (which may not be true) that adding user options to one major mode implies that something needs to be done with the other major mode.

But I think I just had an epiphany -- which is odd because this was my original idea weeks ago but I had forgotten it.

How about having "styles" (for lack of a better term)?  The default style would be called "default".  The default style would try hard to be compatible between ruby-mode and ruby-ts-mode.  Both major modes would need to have the concept of style which was just removed but it isn't hard to add it into ruby-ts-mode.  And I assume it would not be hard to add it into ruby-mode since c-mode has it.

With the default for the style set to default, the two major modes would work the same out of the box as much as possible.

From that point, each mode can have similar or differing styles.  This opens the gate to adding style specific user options.

The reason I "forgot" about this is when I sent my "first draft", I got push back with the styles concept even though c-mode has it.

> I'm not sure what you mean by buffer-local faces.

(Remember, I don't claim to understand Emacs / lisp so this may be false.). My belief is that faces are global.  I can not have a face called foo-face and have it green in one buffer and underline in another buffer.  I've found code on the wiki that seems able to get around this but I didn't test it or try it out since it seemed non-standard.

This sorta comes back to how faces are used currently which I don't grasp.  I would have each major mode have its own set of faces that made sense for that language.  When I added faces in my "first draft" I got questioned "why???" so I gave up on the idea.

> The Bosque del Apache shots from 2001 look great. I hope to have a similar kind of retirement one day.

My advice / suggestion is to take care of yourself.  My eyes, knees, and general health are still pretty good.  I can see well, hike, walk, etc.  I drank drain cleaner when I was 14 months old so I have a lot of internal problems but I'm still able to get around fairly well like I'm 40. 

Take care of your knees, ankles, and feet in particular.  My older bother had one ankle replace last year and is having the other ankle replaced later this month.  Not fun... not fun at all.

Also... time flies.  My last tour at IBM started in 2007.  I moved to another company in 2018.  I have zero idea what happened in those 11 years.  **Poof** they would gone.

**Comment by dgutov** [comment]
> Can you elaborate on this? Is this extra C code in someplace like treesit.c? Or extra lisp code in treesit.el? Or extra code in the modes like ruby-ts-mode.el?

IIUC it will require an additional framework inside treesit.el and some annotations/rules in every ts-based mode. An example of such annotations is here: https://github.com/tree-sitter/tree-sitter-ruby/blob/master/queries/locals.scm, except we'll probably going to write some version of it in Elisp.

> Mostly it is the indent rules that give me pause but I also get the feeling of being tied to a paradigm that is now out dated and also extremely Lisp focused.

Not sure what you mean by Lisp-focused, a lot of the rules are informed by the actual Ruby code in the wild. Most of those authors have never seen Lisp in their life. Although Matz was inspired by Emacs Lisp when he started Ruby, as he later tells it.

> Also, I assume (which may not be true) that adding user options to one major mode implies that something needs to be done with the other major mode.

Something -- yes. When a new behavior is added, we might discuss whether it's feasible and valuable enough to have a common option (inside ruby-mode), or whether a ruby-ts option will suffice. Reimplementing every behavior in ruby-mode yourself won't be a requirement.

> How about having "styles" (for lack of a better term)? The default style would be called "default". The default style would try hard to be compatible between ruby-mode and ruby-ts-mode. Both major modes would need to have the concept of style which was just removed but it isn't hard to add it into ruby-ts-mode. And I assume it would not be hard to add it into ruby-mode since c-mode has it.

I have indeed removed the "styles" thing when checking in the mode, but simply because it was unused. We could still add it. If it is implemented the same was as in `c-ts-mode`, though, it will only work in `ruby-ts-mode`.

Otherwise, a "style" could be thought of as a collection of values for indentation variables. Perhaps this would be useful if we grow many more such variables (CC Mode has lots of them, that's where its styles arose from).

In any case, I'm open to bringing back styles in ruby-ts-mode as long as the patch adding them actually contains alternative styles. What's your goal in this? Do you expect the alternative style to be generally useful, not just to yourself? If yes, let's see the code. If not (or not sure), overriding indentation in `ruby-ts-mode` is as easy as

```rb
(defun my-indent-settings ()
  (let ((rules '(((parent-is "program") parent 0)
                 ...)))
    (setq treesit-simple-indent-rules
          (treesit--indent-rules-optimize rules))))

(add-hook 'ruby-ts-mode-hook #'my-indent-settings)
```

I believe this was what people were referring to in the "push back".

> My belief is that faces are global. I can not have a face called foo-face and have it green in one buffer and underline in another buffer. I've found code on the wiki that seems able to get around this but I didn't test it or try it out since it seemed non-standard.

That is true, pretty much.

> This sorta comes back to how faces are used currently which I don't grasp. I would have each major mode have its own set of faces that made sense for that language. When I added faces in my "first draft" I got questioned "why???" so I gave up on the idea.

Having lots of special faces everywhere will, at the very least, mean that the theme authors will routinely forget to style most of them. It's likely to become impossible to use any theme reliably except the default one. And the default one will require alterations as every new mode is added. That simply doesn't scale.

Major modes (especially third-party ones) may add a few of their own faces for some specific syntax particular for that language, but every time that happens, it basically means the built-in set was insufficient.

Since we're building a set of new major modes, we might as well add some new faces to font-lock.el to do this proper, rather than invent special cases from the get-go (8 new ones got added already for Emacs 29).

> Take care of your knees, ankles, and feet in particular. My older bother had one ankle replace last year and is having the other ankle replaced later this month. Not fun... not fun at all.

> Also... time flies. My last tour at IBM started in 2007. I moved to another company in 2018. I have zero idea what happened in those 11 years. Poof they would gone.

Solid advice. And time flies indeed -- I'm almost a decade into my current employment myself.

**Comment by pedz** [comment]
> Not sure what you mean by Lisp-focused, ...

Not really the indent rules but the general structure of the philosophy is based in lisp -- at least to me.

Let's switch topics a bit.  You bring up themes.

Please go play on this web site: https://paletton.com

I could play on that web site for hours.  I noticed that the guy has changed the UI since my last visit and it may not be as obvious what is going on under the covers as it was before.

I don't use "Themes" because either I don't understand them or they are utterly and completely lost -- and I assume its 50/50 of both.

A theme should be the user picks some small number of colors: 2, 3, or 4 (not the colors themselves -- just a number).  Then he picks a base color and the other colors (not 2 or 3 but roughly 12 to 16) are picked algorithmically based upon a few other parameters.  Like the web site, these two or three other parameters should be tweaked via a UI.

These colors are them loaded into whatever it is that you are designing (Emacs, a web page, etc).  To me, that's how themes should work.

These algorithm are actually ancient -- way before computing was developed.  Artists have known about these principles for ages but they are quickly being forgotten and lost.

About 10 years ago there was a system based upon SASS (the CSS meta language) and it touted itself as being done by designers.  I got all excited and checked into it.  No... they had zero concepts to coordinate the two to four main colors that define the 12 to 16 supporting color scheme.

Also... I don't see it in the UI now but the 12 to 16 colors have priorities.  There are general rules so that if you need, say, only 9 colors, which 9 or the 12 or 16 colors to pick is limited to a few choices.

Colors and coordinating colors has been studied a huge amount and the human preferences and choices are very well understood.  Much like music theory is to music.

So... I don't really constrain myself to worrying about "themes".  They are going about things in a way that makes no sense to me at all.

Here is another example -- although very tangential.

In 1990, I went to work at Tandem.  They were an AT&T shop at the time so they had AT&T tools.  One of those tools was called cscope.  This tool is now public but it wasn't at the time.

You point cscope towards your C code base (hundreds if not thousands of files) and let it crunch for a half a day (at the time) and it spewed out this "database".  With the cscope tool, you could search and find things like all the places that called function foo or all the places that set variable dog.

I connected it up to Emacs so I could click around in the source code and drive cscope to bring up lists of what I was trying to find such as all the places that modified a global variable.

This is 99% of what "LSP" is -- and I was using it back in 1990.

The story continues... if you are not asleep yet

I went back for another tour at IBM and by that point, cscope was available to IBM and so I started showing people the beauty of cscope.  Even without Emacs, it was extremely useful.  1000 times better than grep which is what everyone else was using.  Of course -- practically ZERO people switched.

A few years go by and someone "introduces cscope to IBM" and everyone gets all excited and pats the guy on the back.  Maybe three to five inside friends knew the whole story and just shook their head -- occasionally pointing out that cscope had been introduced to them by me four years before but it did little good.  Well... to be fair, I was already making two to ten times what the guy sitting next to me was making at the time so... it wasn't like my reputation needed bolstering.

And now all the rage if lsp-mode in Emacs.  Yawn... seriously.  A rough approximation of what I had 32 years ago.  32 years is a long freaking time.  (And oh, by the way, especially in the case of Ruby, they are not doing it the way they should be -- but that's a whole other rant.)

I have dozens of stories like these.  Tools I created or processes I developed that I would happily share but practically no one would even have a glimmer of what I was talking about.  There have been few exceptions.

So... when you ask if what I come up with will be useful to others -- my reply is something like: it isn't my fault that it isn't used by others.

**Comment by dgutov** [comment]
> Please go play on this web site: https://paletton.com/

Palette pickers can be very handy to choose the colors, but then that choice needs to be saved and applied to the text on the screen. That implies a particular storage format and a mechanism to associate the colors with tokens on the screen.

I've seen different theme formats, but the considerations I've previously mentioned should apply to all of them. Unless you pick the colors from a palette at random, or using some programmable algorithm, which might work okay-ish, but unlikely to provide optimal results.

> A theme should be the user picks some small number of colors: 2, 3, or 4 (not the colors themselves -- just a number). Then he picks a base color and the other colors (not 2 or 3 but roughly 12 to 16) are picked algorithmically based upon a few other parameters. Like the web site, these two or three other parameters should be tweaked via a UI.

So that's also what you meant, too. It would save some people a lot of hassle, but I believe one couldn't reach optimal results without manual tweaking later. Not all syntactic constructs are equal. Having unrelated ones painted in similar colors (stemming from the same base color) -- but related ones in different ones -- would lead to usability problems.

> This is 99% of what "LSP" is -- and I was using it back in 1990.

I haven't had much occasions to use it (not working in C), but CScope is pretty cool. For a 30-year old tool especially.

LSP might look similar, but its primary achievement is that it created an ecosystem where language tooling authors and editor developers managed to stop reinventing and rewriting language intelligence support all over again, having only N + M codebases, rather than the half-broken always-in-progress N * M. Generalizing the protocol the way it fits many languages (more or less) was a feat as well. CScope meanwhile doesn't even claim proper C++ support.

> I have dozens of stories like these. Tools I created or processes I developed that I would happily share but practically no one would even have a glimmer of what I was talking about. There have been few exceptions.

I could share some stories like that myself. They usually end up with somebody else re-appropriating the idea after you are gone.

What I asked, though, if whether you intend to share the particular style to be used by others. Or whether you just wanted an ability to choose your own indentation style. The latter, as demonstrated, is pretty much available already.

To sum up, I'm happy to facilitate `ruby-ts-mode` in Emacs to contain whatever features you deem necessary. It's up to you, though, whether you want to bear the communication overhead and some bureaucracy, or just develop your own thing independently.

**Comment by pedz** [comment]
> What I asked, though, if whether you intend to share the particular style to be used by others. Or whether you just wanted an ability to choose your own indentation style. The latter, as demonstrated, is pretty much available already.

> To sum up, I'm happy to facilitate ruby-ts-mode in Emacs to contain whatever features you deem necessary. It's up to you, though, whether you want to bear the communication overhead and some bureaucracy, or just develop your own thing independently.

The two statements confuse me.  I hope I answer your questions.  If not, keep asking :-)

If / when I ever get back to this, whatever I come up with will be here.  It might read "Copyright Perry Smith" but I'm happy to share.  I put copyrights mostly out of listening to RMS and somewhere is usually a blurb about, "if you want to use this, please ask".

You are most welcome to come ask me for help but you are pretty bright and I doubt if that will be necessary.

I doubt I will participate much in the developers mailing list.  In the recent past, I post to Reddit Emacs "Hey... I just did XYZ..." and point to a repo but that's about it.

The addition of the local variable stuff interests me but I am not letting myself look at it.  I really need to go out, explore, photograph, etc. or I'm going to be angry with myself and that not good for anyone.

**Comment by dgutov** [comment]
It seems clear enough now. That's okay.

Could you please finalize the copyright assignment stuff, though? At least for the ruby-ts-mode code which is already in Emacs.

**Comment by pedz** [comment]
It's waiting for them.

**Comment by dgutov** [comment]
For the copyright clerk's response? That's fine, then. As long as the process is chugging along.

Thank you.
